pkg/gardenlet/controller/vpaevictionrequirements/reconciler.go

```diff
@@ -0,0 +1,164 @@
+	log := logf.FromContext(seedCtx)
+	if !metav1.HasAnnotation(vpa.ObjectMeta, constants.AnnotationShootMaintenanceWindow) {
+		err := errors.New("didn't find maintenance window annotation on VPA, but VPA had label to be downscaled in maintenance only")
+		return reconcile.Result{}, err
+	}
+
+	windowAnnotation := vpa.GetAnnotations()[constants.AnnotationShootMaintenanceWindow]
+	splitWindowAnnotation := strings.Split(windowAnnotation, ",")
+	maintenanceTimeWindow, err := timewindow.ParseMaintenanceTimeWindow(splitWindowAnnotation[0], splitWindowAnnotation[1])
+	if err != nil {
+		log.Error(err, "Error during parsing the maintenance window from annotation", "begin", splitWindowAnnotation[0], "end", splitWindowAnnotation[1])
+		return reconcile.Result{}, err
+	}
+	isNowInMaintenanceTimeWindow := maintenanceTimeWindow.Contains(r.Clock.Now())
+
+	if isNowInMaintenanceTimeWindow {
+		log.Info("Shoot is inside maintenance window, removing the EvictionRequirement to allow downscaling", "shoot-namespace", vpa.GetNamespace(), "maintenanceWindow", maintenanceTimeWindow)
+		_, err := controllerutils.GetAndCreateOrMergePatch(seedCtx, r.SeedClient, vpa, func() error {
+			if !hasUpscaleOnlyEvictionRequirement(vpa.Spec.UpdatePolicy.EvictionRequirements) {
+				return nil
+			}
+			vpa.Spec.UpdatePolicy.EvictionRequirements = []*vpaautoscalingv1.EvictionRequirement{}
+			return nil
+		})
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// requeue when the maintenance window ends, such that we can add the EvictionRequirement again
+		endTime := maintenanceTimeWindow.AdjustedEnd(r.Clock.Now())
+		requeueAfter := endTime.Sub(r.Clock.Now())
+		log.Info("Requeueing VPA", "requeueAfter", requeueAfter)
+		return reconcile.Result{RequeueAfter: requeueAfter}, nil
+	} else {
+		log.Info("Shoot is not inside maintenance window, adding EvictionRequirement to deny downscaling", "shoot-namespace", vpa.GetNamespace(), "maintenanceWindow", maintenanceTimeWindow)
+		_, err := controllerutils.GetAndCreateOrMergePatch(seedCtx, r.SeedClient, vpa, func() error {
+			if hasUpscaleOnlyEvictionRequirement(vpa.Spec.UpdatePolicy.EvictionRequirements) {
+				return nil
+			}
+			vpa.Spec.UpdatePolicy.EvictionRequirements = []*vpaautoscalingv1.EvictionRequirement{
+				{
+					Resources:         cpuAndMemory,
+					ChangeRequirement: vpaautoscalingv1.TargetHigherThanRequests,
+				},
+			}
+			return nil
+		})
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// requeue when the next maintenance window begins, such that we can remove the EvictionRequirement
+		nextWindowBegin := maintenanceTimeWindow.AdjustedBegin(r.Clock.Now())
+		if nextWindowBegin.Before(r.Clock.Now()) {
+			nextWindowBegin = nextWindowBegin.AddDate(0, 0, 1)
+		}
+		requeueAfter := nextWindowBegin.Sub(r.Clock.Now())
+		log.Info("Requeueing VPA", "requeueAfter", requeueAfter)
+		return reconcile.Result{RequeueAfter: requeueAfter}, nil
```

Comment:
This and also the requeueing above for maintenance end relies on the controller still running at this point in time. What happens if the controller/`gardenlet` crashes/is rescheduled? It seems like we would not see those `vpa` resources again until they are changed again. This might lead to `vpa` objects having eviction requirements during the maintenance window, but also to some having none outside of the maintenance window.

Could you please explain?

Reply:
My understanding is that when the controller is restarted, client-go will end up executing `ListAndWatch`, resulting in an initial list of _all_ VPAs and the controller ends up reconciling all VPAs with the matching label. So the scenario you're describing above should not happen.

When deleting gardenlet, I found indication in the logs that all VPAs were reconciled after the the controller was back, even though the VPAs themselves were not changed:

```
{"level":"info","ts":"2024-01-09T09:12:27.126Z","msg":"Starting workers","controller":"vpa-evictionrequirements","worker count":5}
{"level":"info","ts":"2024-01-09T09:12:27.126Z","msg":"Reconciling vpa","controller":"vpa-evictionrequirements","namespace":"shoot--local--local","name":"etcd-main-vpa","reconcileID":"65523758-44b1-4b4d-aac0-9f7c0af44496","namespace/name":{"name":"etcd-main-vpa","namespace":"shoot
--local--local"}}
{"level":"info","ts":"2024-01-09T09:12:27.126Z","msg":"Found the label evictionrequirements.gardener.cloud/allow-downscale-in-maintenance-window-only","controller":"vpa-evictionrequirements","namespace":"shoot--local--local","name":"etcd-main-vpa","reconcileID":"65523758-44b1-4b4d
-aac0-9f7c0af44496","namespace/name":{"name":"etcd-main-vpa","namespace":"shoot--local--local"}}
{"level":"info","ts":"2024-01-09T09:12:27.126Z","msg":"Shoot is not inside maintenance window, adding EvictionRequirement to deny downscaling","controller":"vpa-evictionrequirements","namespace":"shoot--local--local","name":"etcd-main-vpa","reconcileID":"65523758-44b1-4b4d-aac0-9f
7c0af44496","namespace":"shoot--local--local","name":"etcd-main-vpa","shoot-namespace":"shoot--local--local","maintenanceWindow":"begin=10:00:00, end=11:00:00"}
{"level":"info","ts":"2024-01-09T09:12:27.126Z","msg":"Reconciling vpa","controller":"vpa-evictionrequirements","namespace":"shoot--local--local","name":"etcd-events-vpa","reconcileID":"88b5cb61-2d16-450a-a395-3df77f808937","namespace/name":{"name":"etcd-events-vpa","namespace":"s
hoot--local--local"}}
{"level":"info","ts":"2024-01-09T09:12:27.126Z","msg":"Found the label evictionrequirements.gardener.cloud/allow-downscale-in-maintenance-window-only","controller":"vpa-evictionrequirements","namespace":"shoot--local--local","name":"etcd-events-vpa","reconcileID":"88b5cb61-2d16-45
0a-a395-3df77f808937","namespace/name":{"name":"etcd-events-vpa","namespace":"shoot--local--local"}}
{"level":"info","ts":"2024-01-09T09:12:27.126Z","msg":"Shoot is not inside maintenance window, adding EvictionRequirement to deny downscaling","controller":"vpa-evictionrequirements","namespace":"shoot--local--local","name":"etcd-events-vpa","reconcileID":"88b5cb61-2d16-450a-a395-
3df77f808937","namespace":"shoot--local--local","name":"etcd-events-vpa","shoot-namespace":"shoot--local--local","maintenanceWindow":"begin=10:00:00, end=11:00:00"}
```